subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputUtil.java

```diff
@@ -35,66 +36,106 @@
 import static org.gradle.internal.snapshot.MerkleDirectorySnapshotBuilder.EmptyDirectoryHandlingStrategy.EXCLUDE_EMPTY_DIRS;
 import static org.gradle.internal.snapshot.MerkleDirectorySnapshotBuilder.EmptyDirectoryHandlingStrategy.INCLUDE_EMPTY_DIRS;
+import static org.gradle.internal.snapshot.SnapshotUtil.index;
 
-/**
- * Filters out fingerprints that are not considered outputs. Entries that are considered outputs are:
- * <ul>
- * <li>an entry that did not exist before the execution, but exists after the execution</li>
- * <li>an entry that did exist before the execution, and has been changed during the execution</li>
- * <li>an entry that did wasn't changed during the execution, but was already considered an output during the previous execution</li>
- * </ul>
- */
-public class OutputFilterUtil {
+public class OutputUtil {
 
-    public static ImmutableList<FileSystemSnapshot> filterOutputSnapshotBeforeExecution(FileCollectionFingerprint afterLastExecutionFingerprint, FileSystemSnapshot beforeExecutionOutputSnapshot) {
-        Map<String, FileSystemLocationFingerprint> fingerprints = afterLastExecutionFingerprint.getFingerprints();
+    /**
+     * Filters out snapshots that are not considered outputs. Entries that are considered outputs are:
+     * <ul>
+     * <li>an entry that did not exist before the execution, but exists after the execution</li>
+     * <li>an entry that did exist before the execution, and has been changed during the execution</li>
+     * <li>an entry that did wasn't changed during the execution, but was already considered an output during the previous execution</li>
+     * </ul>
+     */
+    public static ImmutableSortedMap<String, FileSystemSnapshot> filterOutputsWithOverlapBeforeExecution(
+        ImmutableSortedMap<String, FileSystemSnapshot> previousSnapshots,
+        ImmutableSortedMap<String, FileSystemSnapshot> unfilteredBeforeExecutionSnapshots
+    ) {
+        return ImmutableSortedMap.copyOfSorted(
+            Maps.transformEntries(unfilteredBeforeExecutionSnapshots, (key, unfilteredBeforeExecution) -> {
+                    FileSystemSnapshot previous = previousSnapshots.get(key);
+                    if (previous == null) {
+                        return FileSystemSnapshot.EMPTY;
+                    }
+                    //noinspection ConstantConditions
+                    return filterOutputWithOverlapBeforeExecution(previous, unfilteredBeforeExecution);
+                }
+            )
+        );
+    }
+
+    @VisibleForTesting
+    static FileSystemSnapshot filterOutputWithOverlapBeforeExecution(FileSystemSnapshot previous, FileSystemSnapshot beforeExecution) {
+        Map<String, CompleteFileSystemLocationSnapshot> beforeExecutionIndex = index(previous);
         SnapshotFilteringVisitor filteringVisitor = new SnapshotFilteringVisitor((snapshot, isRoot) ->
+            // TODO The first part of this condition is not needed, right?
```

Comment:
That depends on whether missing roots are part of the output snapshot or not.

Reply:
We've removed this together today.